src/backend/distributed/utils/statistics_collection.c

```diff
@@ -114,8 +124,160 @@ CollectBasicUsageStatistics(void)
 
 	CommitTransactionCommand();
 
-	return SendHttpPostJsonRequest(STATS_COLLECTION_HOST "/v1/usage_reports",
-								   fields->data, HTTP_TIMEOUT_SECONDS);
+	success = SendHttpPostJsonRequest(STATS_COLLECTION_HOST "/v1/usage_reports",
+									  fields->data, HTTP_TIMEOUT_SECONDS);
+
+	return success;
+}
+
+
+/* CheckForUpdates queries Citus servers for newer releases of Citus. */
+void
+CheckForUpdates(void)
+{
+	StringInfo url = makeStringInfo();
+	appendStringInfoString(url, STATS_COLLECTION_HOST "/v1/releases/latest?edition=");
+
+	if (!UrlEncode(url, CITUS_EDITION))
+	{
+		ereport(WARNING, (errmsg("url encoding '%s' failed", CITUS_EDITION)));
+	}
+
+	if (!SendHttpGetRequest(url->data, HTTP_TIMEOUT_SECONDS, &CheckForUpdatesCallback))
+	{
+		ereport(WARNING, (errmsg("checking for updates failed")));
+	}
+}
+
+
+/*
+ * CheckForUpdatesCallback receives the response for the request sent by
+ * CheckForUpdates(). It processes the response, and if there is a newer release
+ * of Citus available, logs a LOG message. This function returns 0 if there are
+ * any errors in the received response, which means we didn't consume the data.
+ * Otherwise, it returns (size * count) which means we consumed all of the data.
+ */
+static size_t
+CheckForUpdatesCallback(char *contents, size_t size, size_t count, void *userData)
+{
+	const int32 citusVersionMajor = CITUS_VERSION_NUM / 10000;
+	const int32 citusVersionMinor = (CITUS_VERSION_NUM / 100) % 100;
+	const int32 citusVersionPatch = CITUS_VERSION_NUM % 100;
+	Jsonb *responseJsonb = 0;
+	StringInfo releaseVersion = makeStringInfo();
+	int32 releaseMajor = 0;
+	int32 releaseMinor = 0;
+	int32 releasePatch = 0;
+	char *updateType = NULL;
+
+	StringInfo responseNullTerminated = makeStringInfo();
+	appendBinaryStringInfo(responseNullTerminated, contents, size * count);
+
+	/* jsonb_in can throw errors */
+	PG_TRY();
+	{
+		responseJsonb = DatumGetJsonb(
+			DirectFunctionCall1(jsonb_in, CStringGetDatum(responseNullTerminated->data)));
+	}
+	PG_CATCH();
+	{
+		FlushErrorState();
+		return 0;
+	}
+	PG_END_TRY();
+
+	if (!JsonbFieldStr(responseJsonb, "version", releaseVersion) ||
+		!JsonbFieldInt(responseJsonb, "major", &releaseMajor) ||
+		!JsonbFieldInt(responseJsonb, "minor", &releaseMinor) ||
+		!JsonbFieldInt(responseJsonb, "patch", &releasePatch))
```

Comment:
Since `findJsonbValueFromContainer()` can also be used to check field's existence, I thought returning `bool` for the existence of the field and setting the final parameter for result might make sense. Also, even if the field exists the value can be of incorrect type, so replacing these checks with `jsonb_exists` calls won't work. We could add functions like `JsonbIntFieldExists()`, but its body would be almost the same as a function getting the int field value, so I preferred what I did here.

Reply:
Right. It can also return `false` if `responseJsonb` isn't an object. It can be array, string, number, or null.

Yes, this is called synchronously and returning `0` here triggers the warning in `CheckForUpdates()`. It also triggers the `Sending HTTP request failed.` warning in `PerformHttpRequest()` with error code `"Failed writing received data to disk/application."` which means the callback function couldn't process the data.